pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -63,3 +81,226 @@ func GetgovmomiClient(cfg *VSphereConfig) (*govmomi.Client, error) {
+			}
+		}
+	}
+	return faultMsg
+}
+
+// Get the best fit compatible datastore by free space.
+func GetMostFreeDatastore(dsMo []mo.Datastore) mo.Datastore {
+	var curMax int64
+	curMax = -1
+	var index int
+	for i, ds := range dsMo {
+		dsFreeSpace := ds.Info.GetDatastoreInfo().FreeSpace
+		if dsFreeSpace > curMax {
+			curMax = dsFreeSpace
+			index = i
+		}
+	}
+	return dsMo[index]
+}
+
+func (vs *VSphere) GetCompatibleDatastoresMo(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult) ([]mo.Datastore, error) {
+	compatibleHubs := compatibilityResult.CompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(compatibleHubs) < 1 {
+		return nil, fmt.Errorf("There are no compatible datastores that satisfy the storage policy requirements")
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, compatibleHubs)
+	if err != nil {
+		return nil, err
+	}
+	return dsMoList, nil
+}
+
+func (vs *VSphere) GetNonCompatibleDatastoresMo(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult) []mo.Datastore {
+	nonCompatibleHubs := compatibilityResult.NonCompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(nonCompatibleHubs) < 1 {
+		return nil
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, nonCompatibleHubs)
+	if err != nil {
+		return nil
+	}
+	return dsMoList
+}
+
+// Get the datastore managed objects for the place hubs using property collector.
+func (vs *VSphere) getDatastoreMo(ctx context.Context, hubs []pbmtypes.PbmPlacementHub) ([]mo.Datastore, error) {
+	var dsMoRefs []types.ManagedObjectReference
+	for _, hub := range hubs {
+		dsMoRefs = append(dsMoRefs, types.ManagedObjectReference{
+			Type:  hub.HubType,
+			Value: hub.HubId,
+		})
+	}
+
+	pc := property.DefaultCollector(vs.client.Client)
+	var dsMoList []mo.Datastore
+	err := pc.Retrieve(ctx, dsMoRefs, []string{DatastoreInfoProperty}, &dsMoList)
+	if err != nil {
+		return nil, err
+	}
+	return dsMoList, nil
+}
+
+// Get all datastores accessible for the virtual machine object.
+func (vs *VSphere) getSharedDatastoresInK8SCluster(ctx context.Context) ([]types.ManagedObjectReference, error) {
+	f := find.NewFinder(vs.client.Client, true)
+	dc, err := f.Datacenter(ctx, vs.cfg.Global.Datacenter)
+	f.SetDatacenter(dc)
+	vmFolder, err := f.Folder(ctx, strings.TrimSuffix(vs.cfg.Global.WorkingDir, "/"))
+	if err != nil {
+		return nil, err
+	}
+	vmMoList, err := vs.GetVMsInsideFolder(ctx, vmFolder, []string{NameProperty})
+	if err != nil {
+		return nil, err
+	}
+	index := 0
+	var sharedDs []string
+	for _, vmMo := range vmMoList {
+		if !strings.HasPrefix(vmMo.Name, DummyVMPrefixName) {
+			accessibleDatastores, err := vs.getAllAccessibleDatastores(ctx, vmMo)
+			if err != nil {
+				return nil, err
+			}
+			if index == 0 {
+				sharedDs = accessibleDatastores
+			} else {
+				sharedDs = intersect(sharedDs, accessibleDatastores)
+				if len(sharedDs) == 0 {
```

Comment:
No, if the intersect of 2 lists is 0. Then we should return immediately as further intersects would always result in 0 values.